# Financial data analysis

Financial data analysis involves examining financial information, such as stock prices, market trends, and company performance, to derive insights that support decision-making. We analyze metrics like volatility, returns, and various risk assessment methods. In this article, I’ll walk you through financial data analysis with Python, which will help you understand how to analyze financial data and make decisions based on it.

## Ananlysis goal:

This analysis aims to explore financial data from NIFTY50 stocks to uncover insights that can guide investment strategies and risk management decisions. The dataset consists of 24 days of historical closing prices for 50 stocks, with the Date column representing trading days.

## Analysis scope:

The scope of the analysis includes calculating descriptive statistics to summarize stock behaviour, constructing and evaluating a portfolio for returns and risk, assessing volatility and Value at Risk (VaR), identifying trends through technical indicators like moving averages and Bollinger Bands, and forecasting future stock prices using Monte Carlo simulations.

In [1]:
# importing required library

import pandas as pd

In [2]:
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

In [3]:
# importing the dataset
df = pd.read_csv("data/nifty50_closing_prices.csv")
df.head()

Date  RELIANCE.NS  HDFCBANK.NS  ICICIBANK.NS  \
0  2024-08-20 00:00:00+05:30  2991.899902  1637.699951   1179.449951   
1  2024-08-21 00:00:00+05:30  2997.350098  1625.800049   1174.849976   
2  2024-08-22 00:00:00+05:30  2996.250000  1631.300049   1191.099976   
3  2024-08-23 00:00:00+05:30  2999.949951  1625.050049   1203.500000   
4  2024-08-26 00:00:00+05:30  3025.199951  1639.949951   1213.300049   

       INFY.NS       TCS.NS  KOTAKBANK.NS  HINDUNILVR.NS      ITC.NS  \
0  1872.199951  4523.299805   1805.650024    2751.050049  498.799988   
1  1872.699951  4551.500000   1812.949951    2791.199951  505.399994   
2  1880.250000  4502.000000   1821.500000    2792.800049  504.549988   
3  1862.099976  4463.899902   1818.000000    2815.600098  505.799988   
4  1876.150024  4502.450195   1812.500000    2821.149902  505.700012   

         LT.NS     SBIN.NS  BAJFINANCE.NS  BHARTIARTL.NS   HCLTECH.NS  \
0  3572.699951  820.299988    6722.200195    1449.150024  1686.750000   
1  3596.050049  815.549988    6735.350098    1463.449951  1677.250000   
2  3606.500000  820.299988    6743.600098    1486.349976  1676.150024   
3  3598.550049  815.349976    6735.850098    1506.750000  1661.449951   
4  3641.899902  815.049988    6778.350098    1513.550049  1719.449951   

   ASIANPAINT.NS  AXISBANK.NS     DMART.NS     MARUTI.NS  ULTRACEMCO.NS  \
0    3103.199951  1168.000000  5079.200195  12214.950195   11349.700195   
1    3151.550049  1174.400024  5099.450195  12220.950195   11200.900391   
2    3186.600098  1169.949951  5057.850098  12276.349609   11309.400391   
3    3154.649902  1165.949951  4901.500000  12302.299805   11341.799805   
4    3171.350098  1170.300049  4959.399902  12243.799805   11337.099609   

   HDFC.NS     TITAN.NS  SUNPHARMA.NS       M&M.NS  NESTLEIND.NS    WIPRO.NS  \
0      NaN  3474.899902   1766.349976  2771.300049   2518.500000  524.650024   
1      NaN  3560.399902   1764.650024  2769.399902   2551.750000  526.349976   
2      NaN  3604.399902   1750.650024  2732.949951   2551.000000  519.000000   
3      NaN  3570.000000   1775.750000  2759.000000   2529.199951  512.400024   
4      NaN  3630.199951   1772.449951  2793.100098   2519.550049  520.000000   

   ADANIGREEN.NS  TATASTEEL.NS  JSWSTEEL.NS  POWERGRID.NS     ONGC.NS  \
0    1924.599976    153.929993   917.150024    340.500000  327.555695   
1    1920.849976    151.919998   925.799988    336.649994  325.174194   
2    1886.349976    154.139999   933.250000    334.000000  321.850006   
3    1900.900024    154.199997   941.049988    336.250000  318.899994   
4    1888.900024    155.699997   963.500000    338.250000  327.850006   

      NTPC.NS  COALINDIA.NS     BPCL.NS      IOC.NS     TECHM.NS  \
0  406.250000    524.599976  349.399994  172.229996  1628.599976   
1  408.950012    532.200012  351.200012  173.889999  1604.650024   
2  403.350006    528.849976  350.100006  173.789993  1611.250000   
3  401.950012    538.849976  352.200012  173.130005  1598.400024   
4  414.850006    538.099976  351.149994  173.460007  1640.150024   

   INDUSINDBK.NS  DIVISLAB.NS    GRASIM.NS     CIPLA.NS  BAJAJFINSV.NS  \
0    1381.300049  4723.149902  2636.699951  1562.849976    1602.099976   
1    1384.000000  4900.799805  2684.850098  1594.599976    1620.949951   
2    1381.900024  4911.450195  2755.149902  1585.800049    1625.699951   
3    1388.550049  4855.950195  2748.550049  1574.550049    1639.900024   
4    1384.500000  4926.250000  2736.600098  1593.949951    1686.199951   

   TATAMOTORS.NS  HEROMOTOCO.NS   DRREDDY.NS   SHREECEM.NS  BRITANNIA.NS  \
0    1086.900024    5244.399902  6965.350098  24730.550781   5765.799805   
1    1085.199951    5284.700195  7062.450195  24808.050781   5837.350098   
2    1068.449951    5329.950195  6969.049805  25012.400391   5836.799805   
3    1085.150024    5384.899902  6954.500000  24706.050781   5792.649902   
4    1092.400024    5343.750000  6943.299805  24906.449219   5796.950195   

       UPL.NS  EICHERMOT.NS   SBILIFE

## Data preprocessing 

In [4]:
# checking missing values
missing_values = df.isnull().sum()
missing_values

Date              0
RELIANCE.NS       0
HDFCBANK.NS       0
ICICIBANK.NS      0
INFY.NS           0
TCS.NS            0
KOTAKBANK.NS      0
HINDUNILVR.NS     0
ITC.NS            0
LT.NS             0
SBIN.NS           0
BAJFINANCE.NS     0
BHARTIARTL.NS     0
HCLTECH.NS        0
ASIANPAINT.NS     0
AXISBANK.NS       0
DMART.NS          0
MARUTI.NS         0
ULTRACEMCO.NS     0
HDFC.NS          24
TITAN.NS          0
SUNPHARMA.NS      0
M&M.NS            0
NESTLEIND.NS      0
WIPRO.NS          0
ADANIGREEN.NS     0
TATASTEEL.NS      0
JSWSTEEL.NS       0
POWERGRID.NS      0
ONGC.NS           0
NTPC.NS           0
COALINDIA.NS      0
BPCL.NS           0
IOC.NS            0
TECHM.NS          0
INDUSINDBK.NS     0
DIVISLAB.NS       0
GRASIM.NS         0
CIPLA.NS          0
BAJAJFINSV.NS     0
TATAMOTORS.NS     0
HEROMOTOCO.NS     0
DRREDDY.NS        0
SHREECEM.NS       0
BRITANNIA.NS      0
UPL.NS            0
EICHERMOT.NS      0
SBILIFE.NS        0
ADANIPORTS.NS     0
BAJAJ-AUTO.NS     0


In [5]:
# check for date column format
date_format_check = pd.to_datetime(df['Date'], errors='coerce').notna().all()
date_format_check

True

In [6]:
# check if data has sufficient rows for time-series analysis(at least 20 rows)
sufficient_rows = df.shape[0] >= 20
sufficient_rows

True

In [7]:
# preparing a summary of the checks 
data_preparation_status = {
    "Missing Values in columns": missing_values[missing_values > 0].to_dict(),
    "Date Column format valid": date_format_check,
    "Sufficient rows for time-series analysis": sufficient_rows
}

data_preparation_status


{'Missing Values in columns': {'HDFC.NS': 24},
 'Date Column format valid': True,
 'Sufficient rows for time-series analysis': True}

## Data preparation

In [8]:
# dropping the HDFC.NS column
df = df.drop(columns=['HDFC.NS'])

In [11]:
# Converting the Date column to datetime format
df['Date'] = pd.to_datetime(df['Date'])

df['Date'].dtype

datetime64[ns, UTC+05:30]

In [12]:
# sorting the dataset by date to ensure proper time-series order
df = df.sort_values(by="Date")

In [13]:
# reset index for a clean dataframe
df.reset_index(drop=True, inplace=True)

## Descriptive statistics


In [14]:
descriptive_stats = df.describe().T # Transpose for better readability
descriptive_stats

count          mean         std           min           25%  \
RELIANCE.NS     24.0   2976.912506   41.290551   2903.000000   2941.862488   
HDFCBANK.NS     24.0   1652.339579   28.258220   1625.050049   1637.062469   
ICICIBANK.NS    24.0   1236.770818   36.438726   1174.849976   1219.750031   
INFY.NS         24.0   1914.558324   30.240685   1862.099976   1893.687469   
TCS.NS          24.0   4478.349976   70.822718   4284.899902   4472.937622   
KOTAKBANK.NS    24.0   1809.422918   32.936318   1764.150024   1783.800049   
HINDUNILVR.NS   24.0   2845.333344   65.620694   2751.050049   2790.662476   
ITC.NS          24.0    507.739581    5.472559    497.299988    504.962502   
LT.NS           24.0   3647.099976   60.511574   3536.949951   3597.950012   
SBIN.NS         24.0    802.233332   17.442330    768.599976    785.224991   
BAJFINANCE.NS   24.0   7203.118754  306.658594   6722.200195   6890.900024   
BHARTIARTL.NS   24.0   1572.574997   67.346274   1449.150024   1535.125000   
HCLTECH.NS      24.0   1753.743744   46.874886   1661.449951   1719.449951   
ASIANPAINT.NS   24.0   3231.654175   88.793647   3103.199951   3153.874939   
AXISBANK.NS     24.0   1191.879155   27.369408   1158.750000   1170.924957   
DMART.NS        24.0   5143.058329  155.593701   4901.500000   4995.037598   
MARUTI.NS       24.0  12320.356201  109.587342  12145.750000  12243.499756   
ULTRACEMCO.NS   24.0  11472.318807  172.673053  11200.900391  11330.174805   
TITAN.NS        24.0   3654.899974   95.697721   3474.899902   3568.787476   
SUNPHARMA.NS    24.0   1819.299993   34.792913   1750.650024   1796.749969   
M&M.NS          24.0   2763.954183   56.045817   2654.250000   2737.562561   
NESTLEIND.NS    24.0   2539.102081   46.123738   2492.500000   2515.037476   
WIPRO.NS        24.0    529.764582   11.824190    512.400024    519.787506   
ADANIGREEN.NS   24.0   1891.595835   54.031206   1788.199951   1858.249969   
TATASTEEL.NS    24.0    152.277083    1.893183    148.169998    151.209999   
JSWSTEEL.NS     24.0    943.729167   15.778456    917.150024    932.987488   
POWERGRID.NS    24.0    335.285414    3.013865    328.549988    333.824997   
ONGC.NS         24.0    309.819995   16.989364    285.250000    293.662491   
NTPC.NS         24.0    407.133334    8.990767    389.649994    402.925003   
COALINDIA.NS    24.0    507.735413   20.470753    477.950012    489.824997   
BPCL.NS         24.0    347.529167    9.011248    324.450012    341.887489   
IOC.NS          24.0    173.630416    3.702380    165.039993    172.127499   
TECHM.NS        24.0   1626.229172   21.236330   1579.199951   1606.900024   
INDUSINDBK.NS   24.0   1428.679164   33.914618   1381.300049   1404.149994   
DIVISLAB.NS     24.0   5171.531250  247.674895   4723.149902   4990.974854   
GRASIM.NS       24.0   2718.235443   35.912080   2636.699951   2692.337585   
CIPLA.NS        24.0   1627.025004   29.773691   1562.849976   1607.800049   
BAJAJFINSV.NS   24.0   1796.470825   99.422795   1602.099976   1717.625000   
TATAMOTORS.NS   24.0   1044.662498   52.496391    962.049988    987.837524   
HEROMOTOCO.NS   24.0   5619.377096  247.092728   5244.399902   5370.012451   
DRREDDY.NS      24.0   6785.795817  175.124908   6502.549805   6649.574951   
SHREECEM.NS     24.0  25299.906169  429.919834  24692.199219  24885.174316   
BRITANNIA.NS    24.0   5935.202026  144.164343   5703.350098   5835.449829   
UPL.NS          24.0    596.343750   16.975821    566.150024    578.875015   
EICHERMOT.NS    24.0   4863.831258   68.442418   4726.649902   4846.037598   
SBILIFE.NS      24.0   1849.331243   43.189734   1761.300049   1820.725006   
ADANIPORTS.NS   24.0   1462.916677   26.223794   1408.199951   1440.199951   
BAJAJ-AUTO.NS   24.0  10999.654134  659.810841   9779.700195  10617.962402   
HINDALCO.NS     24.0    681.885417   15.952804    647.700012    672.162521   

                        50%           75%           max  
RELIANCE.NS     2988.924927   3005.237427   3041.850098 

In [15]:
descriptive_stats = descriptive_stats[['mean', 'std', 'min', 'max']]
descriptive_stats

mean         std           min           max
RELIANCE.NS     2976.912506   41.290551   2903.000000   3041.850098
HDFCBANK.NS     1652.339579   28.258220   1625.050049   1741.199951
ICICIBANK.NS    1236.770818   36.438726   1174.849976   1338.449951
INFY.NS         1914.558324   30.240685   1862.099976   1964.500000
TCS.NS          4478.349976   70.822718   4284.899902   4553.750000
KOTAKBANK.NS    1809.422918   32.936318   1764.150024   1904.500000
HINDUNILVR.NS   2845.333344   65.620694   2751.050049   2977.600098
ITC.NS           507.739581    5.472559    497.299988    519.500000
LT.NS           3647.099976   60.511574   3536.949951   3793.899902
SBIN.NS          802.233332   17.442330    768.599976    824.799988
BAJFINANCE.NS   7203.118754  306.658594   6722.200195   7631.100098
BHARTIARTL.NS   1572.574997   67.346274   1449.150024   1711.750000
HCLTECH.NS      1753.743744   46.874886   1661.449951   1813.750000
ASIANPAINT.NS   3231.654175   88.793647   3103.199951   3383.250000
AXISBANK.NS     1191.879155   27.369408   1158.750000   1245.000000
DMART.NS        5143.058329  155.593701   4901.500000   5361.399902
MARUTI.NS      12320.356201  109.587342  12145.750000  12614.500000
ULTRACEMCO.NS  11472.318807  172.673053  11200.900391  11798.299805
TITAN.NS        3654.899974   95.697721   3474.899902   3797.199951
SUNPHARMA.NS    1819.299993   34.792913   1750.650024   1866.099976
M&M.NS          2763.954183   56.045817   2654.250000   2950.850098
NESTLEIND.NS    2539.102081   46.123738   2492.500000   2699.550049
WIPRO.NS         529.764582   11.824190    512.400024    551.900024
ADANIGREEN.NS   1891.595835   54.031206   1788.199951   2003.949951
TATASTEEL.NS     152.277083    1.893183    148.169998    155.699997
JSWSTEEL.NS      943.729167   15.778456    917.150024    981.549988
POWERGRID.NS     335.285414    3.013865    328.549988    340.850006
ONGC.NS          309.819995   16.989364    285.250000    330.750000
NTPC.NS          407.133334    8.990767    389.649994    423.950012
COALINDIA.NS     507.735413   20.470753    477.950012    538.849976
BPCL.NS          347.529167    9.011248    324.450012    360.700012
IOC.NS           173.630416    3.702380    165.039993    181.339996
TECHM.NS        1626.229172   21.236330   1579.199951   1656.050049
INDUSINDBK.NS   1428.679164   33.914618   1381.300049   1484.750000
DIVISLAB.NS     5171.531250  247.674895   4723.149902   5498.649902
GRASIM.NS       2718.235443   35.912080   2636.699951   2784.350098
CIPLA.NS        1627.025004   29.773691   1562.849976   1671.800049
BAJAJFINSV.NS   1796.470825   99.422795   1602.099976   1916.800049
TATAMOTORS.NS   1044.662498   52.496391    962.049988   1121.650024
HEROMOTOCO.NS   5619.377096  247.092728   5244.399902   6013.250000
DRREDDY.NS      6785.795817  175.124908   6502.549805   7062.450195
SHREECEM.NS    25299.906169  429.919834  24692.199219  26019.650391
BRITANNIA.NS    5935.202026  144.164343   5703.350098   6210.549805
UPL.NS           596.343750   16.975821    566.150024    619.200012
EICHERMOT.NS    4863.831258   68.442418   4726.649902   4963.149902
SBILIFE.NS      1849.331243   43.189734   1761.300049   1928.650024
ADANIPORTS.NS   1462.916677   26.223794   1408.199951   1503.500000
BAJAJ-AUTO.NS  10999.654134  659.810841   9779.700195  11950.299805
HINDALCO.NS      681.885417   15.952804    647.700012    711.849976

In [16]:
descriptive_stats.columns = ['Mean', 'Std Dev', 'Min', 'Max']
descriptive_stats

Mean     Std Dev           Min           Max
RELIANCE.NS     2976.912506   41.290551   2903.000000   3041.850098
HDFCBANK.NS     1652.339579   28.258220   1625.050049   1741.199951
ICICIBANK.NS    1236.770818   36.438726   1174.849976   1338.449951
INFY.NS         1914.558324   30.240685   1862.099976   1964.500000
TCS.NS          4478.349976   70.822718   4284.899902   4553.750000
KOTAKBANK.NS    1809.422918   32.936318   1764.150024   1904.500000
HINDUNILVR.NS   2845.333344   65.620694   2751.050049   2977.600098
ITC.NS           507.739581    5.472559    497.299988    519.500000
LT.NS           3647.099976   60.511574   3536.949951   3793.899902
SBIN.NS          802.233332   17.442330    768.599976    824.799988
BAJFINANCE.NS   7203.118754  306.658594   6722.200195   7631.100098
BHARTIARTL.NS   1572.574997   67.346274   1449.150024   1711.750000
HCLTECH.NS      1753.743744   46.874886   1661.449951   1813.750000
ASIANPAINT.NS   3231.654175   88.793647   3103.199951   3383.250000
AXISBANK.NS     1191.879155   27.369408   1158.750000   1245.000000
DMART.NS        5143.058329  155.593701   4901.500000   5361.399902
MARUTI.NS      12320.356201  109.587342  12145.750000  12614.500000
ULTRACEMCO.NS  11472.318807  172.673053  11200.900391  11798.299805
TITAN.NS        3654.899974   95.697721   3474.899902   3797.199951
SUNPHARMA.NS    1819.299993   34.792913   1750.650024   1866.099976
M&M.NS          2763.954183   56.045817   2654.250000   2950.850098
NESTLEIND.NS    2539.102081   46.123738   2492.500000   2699.550049
WIPRO.NS         529.764582   11.824190    512.400024    551.900024
ADANIGREEN.NS   1891.595835   54.031206   1788.199951   2003.949951
TATASTEEL.NS     152.277083    1.893183    148.169998    155.699997
JSWSTEEL.NS      943.729167   15.778456    917.150024    981.549988
POWERGRID.NS     335.285414    3.013865    328.549988    340.850006
ONGC.NS          309.819995   16.989364    285.250000    330.750000
NTPC.NS          407.133334    8.990767    389.649994    423.950012
COALINDIA.NS     507.735413   20.470753    477.950012    538.849976
BPCL.NS          347.529167    9.011248    324.450012    360.700012
IOC.NS           173.630416    3.702380    165.039993    181.339996
TECHM.NS        1626.229172   21.236330   1579.199951   1656.050049
INDUSINDBK.NS   1428.679164   33.914618   1381.300049   1484.750000
DIVISLAB.NS     5171.531250  247.674895   4723.149902   5498.649902
GRASIM.NS       2718.235443   35.912080   2636.699951   2784.350098
CIPLA.NS        1627.025004   29.773691   1562.849976   1671.800049
BAJAJFINSV.NS   1796.470825   99.422795   1602.099976   1916.800049
TATAMOTORS.NS   1044.662498   52.496391    962.049988   1121.650024
HEROMOTOCO.NS   5619.377096  247.092728   5244.399902   6013.250000
DRREDDY.NS      6785.795817  175.124908   6502.549805   7062.450195
SHREECEM.NS    25299.906169  429.919834  24692.199219  26019.650391
BRITANNIA.NS    5935.202026  144.164343   5703.350098   6210.549805
UPL.NS           596.343750   16.975821    566.150024    619.200012
EICHERMOT.NS    4863.831258   68.442418   4726.649902   4963.149902
SBILIFE.NS      1849.331243   43.189734   1761.300049   1928.650024
ADANIPORTS.NS   1462.916677   26.223794   1408.199951   1503.500000
BAJAJ-AUTO.NS  10999.654134  659.810841   9779.700195  11950.299805
HINDALCO.NS      681.885417   15.952804    647.700012    711.849976

## Portfolio analysis

Portfolio Analysis is the process of evaluating the performance of a collection of financial assets (a portfolio) to understand its returns, risks, and overall behaviour. It helps investors optimize asset allocation to achieve specific financial goals. Let’s perform a portfolio analysis:

In [17]:
# assign weights to a subset of stocks (example: RELIANCE.NS, HDFCBANK.NS, ICICIBANK.NS)
weights = [0.4, 0.35, 0.25]
portfolio_data = df[['RELIANCE.NS', 'HDFCBANK.NS', 'ICICIBANK.NS']]
portfolio_data

RELIANCE.NS  HDFCBANK.NS  ICICIBANK.NS
0   2991.899902  1637.699951   1179.449951
1   2997.350098  1625.800049   1174.849976
2   2996.250000  1631.300049   1191.099976
3   2999.949951  1625.050049   1203.500000
4   3025.199951  1639.949951   1213.300049
5   3000.899902  1637.750000   1226.349976
6   2996.600098  1637.099976   1223.849976
7   3041.850098  1638.550049   1221.900024
8   3019.250000  1636.900024   1229.199951
9   3032.500000  1626.949951   1229.949951
10  3018.250000  1637.349976   1247.699951
11  3029.100098  1641.800049   1236.349976
12  2985.949951  1645.449951   1235.949951
13  2929.649902  1636.949951   1208.150024
14  2924.899902  1646.500000   1235.000000
15  2923.050049  1650.349976   1237.300049
16  2903.000000  1643.900024   1236.349976
17  2959.600098  1666.599976   1252.150024
18  2945.250000  1665.949951   1250.349976
19  2942.699951  1670.949951   1262.849976
20  2944.600098  1668.800049   1268.099976
21  2926.899902  1694.800049   1288.349976
22  2939.350098  1708.500000   1292.000000
23  2971.850098  1741.199951   1338.449951

In [19]:
# calculate daily returns
daily_returns = portfolio_data.pct_change().dropna()
daily_returns

RELIANCE.NS  HDFCBANK.NS  ICICIBANK.NS
1      0.001822    -0.007266     -0.003900
2     -0.000367     0.003383      0.013832
3      0.001235    -0.003831      0.010411
4      0.008417     0.009169      0.008143
5     -0.008033    -0.001341      0.010756
6     -0.001433    -0.000397     -0.002039
7      0.015100     0.000886     -0.001593
8     -0.007430    -0.001007      0.005974
9      0.004389    -0.006079      0.000610
10    -0.004699     0.006392      0.014431
11     0.003595     0.002718     -0.009097
12    -0.014245     0.002223     -0.000324
13    -0.018855    -0.005166     -0.022493
14    -0.001621     0.005834      0.022224
15    -0.000632     0.002338      0.001862
16    -0.006859    -0.003908     -0.000768
17     0.019497     0.013809      0.012780
18    -0.004849    -0.000390     -0.001438
19    -0.000866     0.003001      0.009997
20     0.000646    -0.001287      0.004157
21    -0.006011     0.015580      0.015969
22     0.004254     0.008084      0.002833
23     0.011057     0.019140      0.035952

In [21]:
# calculate portfolio returns
portfolio_returns  = (daily_returns * weights).sum(axis=1)
portfolio_returns.head()

1   -0.002790
2    0.004495
3    0.001756
4    0.008612
5   -0.000994
dtype: float64

### Output intrerpretation:

In the output, each value represents the percentage change in the portfolio’s value for a particular day. For example, a return of -0.002790 on the first day indicates a 0.279% decrease in the portfolio’s value, while 0.004495 on the second day indicates a 0.4495% increase. These values help in tracking the portfolio’s daily performance over time.

## Risk assessment


Risk Assessment is the process of evaluating the potential risks in an investment, such as price volatility and potential losses, to help investors make informed decisions.

In [22]:
# Calculate standard deviation(volatility)

volatility = daily_returns.std()
volatility

RELIANCE.NS     0.008708
HDFCBANK.NS     0.006901
ICICIBANK.NS    0.011594
dtype: float64

In [23]:
# Calculate VaR(95% confidence_level)
confidence_level = 0.05
VaR = daily_returns.quantile(confidence_level)
VaR

RELIANCE.NS    -0.013624
HDFCBANK.NS    -0.005987
ICICIBANK.NS   -0.008577
Name: 0.05, dtype: float64

In [24]:
# Display risk metrics
risks_metrics = pd.DataFrame({ "Volatitlity (Std Dev)": volatility, "Value at Risk(VaR)": VaR })
risks_metrics

Volatitlity (Std Dev)  Value at Risk(VaR)
RELIANCE.NS                0.008708           -0.013624
HDFCBANK.NS                0.006901           -0.005987
ICICIBANK.NS               0.011594           -0.008577

### Explanation: 

To perform risk assessment, we:

    - Calculated the standard deviation of daily returns for each stock, to measure how much the stock prices fluctuate.
    - Computed the 5th percentile (95% confidence level) of daily returns, to estimate the maximum loss the portfolio could experience on a bad day.

### Result intrerpretation

The results show the risk metrics for three stocks in the portfolio:

1. **Volatility (Std Dev):** RELIANCE has a volatility of 0.87%, HDFCBANK has 0.69%, and ICICIBANK has 1.16%. This indicates that ICICIBANK has the highest price fluctuations, while HDFCBANK is the least volatile.
2. **Value at Risk (VaR):** At a 95% confidence level, RELIANCE has a maximum potential daily loss of 1.36%, HDFCBANK has 0.60%, and ICICIBANK has 0.86%. These values indicate the risk of loss for each stock in a single day under normal market conditions.